tests/puzzle/test_provider.py

```diff
@@ -113,7 +113,7 @@ def test_search_for(
             Dependency("foo", ">=2"),
             URLDependency("foo", SOME_URL),
             [Package("foo", "3")],
-            [],
+            [Package("foo", "3")],
```

Comment:
Looks like correct behaviour to me: the dependency is not satisfied by the URL package and so it falls back to checking the repository (which exactly the proposed change)